core/src/main/java/org/apache/iceberg/ManifestFilterManager.java

```diff
@@ -289,13 +321,38 @@ private void invalidateFilteredCache() {
     cleanUncommitted(SnapshotProducer.EMPTY_SET);
   }
 
+  private void recordPartitionMinDataSequenceNumber(ManifestFile manifest) {
```

Comment:
The existing approach only looks at manifest metadata to understand the min data sequence number across all partitions. This is really cheap as we don't have to open manifests (which can be a really expensive operation). That leads to the problem if one partition is significantly behind, it prevents garbage collection of delete files in other partitions. We have solved that for position deletes via the `rewritePositionDeletes` action but it still remains open for equality deletes.

I am not convinced opening these manifests during commits is a good idea. Can we explore the option of leveraging the partition stats spec added recently? We are still building an action to generate those stats but let's think through whether it can help us. One option can be to check if the partition stats file is present and use it populate the min data sequence numbers, opening just a single Parquet file vs potentially tons of manifests.

Reply:
Thoughts, @szehon-ho @ajantha-bhat @zinking?

Reply:
> Can we explore the option of leveraging the partition stats spec added

Wow. That sounds like a nice usecase 👍 
For each partition, we do keep the snapshot id that last updated that partition. Using the snapshot id we can extract the data sequence numbers from the snapshot. 

@zinking: The current status of the partition stats project can be tracked from this: https://github.com/apache/iceberg/issues/8450

Another alternative approach is to convert equality delete to position delete (this work is pending), So we can reuse the `rewritePositionDeletes`. But it is a long route.

Reply:
Agree with above.  I think there were some other attempts to do this before too, but the concern here is that you dont want to do a lot of things in the commit critical path (here potentially opening an unlimited number of manifest files).  Yea if its something cheaper to do like reading one partition stats file, it may be better.  Also yes the plan has always been to implement convert eq-delete to pos deletes (which can then be cleaned up by rewritePositionDeletes), though not sure if any progress is being made there.

Reply:
agree that adding this on the general write path is too heavy, so that's why I prefer it enabled during rewrite, or probably just some of the rewrites.

on the other hand, it sounds reasonable to track this on partition metadata, but there has to be somewhere to calculate it some where anyways.

Reply:
re-visited the spec, one more thing is that partitions stats sounds optional, but this is a real issue that needs to be addressed.

Reply:
True. It is optional. 

We can probably have both the ways implemented. If stats available, use it. Else read manifests. Lets see what others think.

Reply:
I would love to have dangling deletes to be removed as part of rewrite operation itself. I think this PR also does the same but not in a distributed way. 

+1 for making it distributed.

Reply:
> I had implemented a version of this as a Spark job. I had thought at the time to integrate it as RewriteAction, something like a finally action after all the compact jobs are run. It may be better as its Spark job and more scalable. #6581 @aokolnychyi should we proceed with it?

seems solving the issue, it's just there are too many such kind of actions, I'm really not a big fan of many separate spark actions. I remember there's another one that removed pos delete etc. can we just put these in rewrite so that the end users can be assured that after rewrite everything returns to normal?

Reply:
I think what Szehon meant is to call the cleanup action from the finally block of rewrite action. Not expose as a new action for cleanup.

Reply:
I'll need to think about this a bit more but I do like the idea of using partition stats in one or another way. I'll get back next week.

Reply:
Yea for me @zinking raised the good question about using partition stats.  Its optional so if the user hasn't analyzed the table, it will be different behavior.  Which may not be so obvious to users.

Reply:
Okay, I think we all agree that we should use partition stats if they are available and read manifests otherwise. We may think about extending our regular writes to check if there is a partition stats file available and drop the delete files per partition rather than globally, like it is done today. We shouldn't open manifests during writes. We can only do that in a distributed fashion, meaning it has to be part of an action. There we either can add a new action or integrate this logic into the existing action.

Reply:
@szehon-ho and I talked a little bit about this offline. I think we can try extending the action for rewriting data files to also attempt to remove dangling deletes in partitions that were successfully compacted. Separately, we may integrate the cleanup using partition stats during regular commits under a flag (off by default).